src/commonMain/kotlin/fr/acinq/lightning/wire/OfferTypes.kt

```diff
@@ -0,0 +1,1003 @@
+                    OfferCurrency.tag to OfferCurrency as TlvValueReader<InvoiceRequestTlv>,
+                    OfferAmount.tag to OfferAmount as TlvValueReader<InvoiceRequestTlv>,
+                    OfferDescription.tag to OfferDescription as TlvValueReader<InvoiceRequestTlv>,
+                    OfferFeatures.tag to OfferFeatures as TlvValueReader<InvoiceRequestTlv>,
+                    OfferAbsoluteExpiry.tag to OfferAbsoluteExpiry as TlvValueReader<InvoiceRequestTlv>,
+                    OfferPaths.tag to OfferPaths as TlvValueReader<InvoiceRequestTlv>,
+                    OfferIssuer.tag to OfferIssuer as TlvValueReader<InvoiceRequestTlv>,
+                    OfferQuantityMax.tag to OfferQuantityMax as TlvValueReader<InvoiceRequestTlv>,
+                    OfferNodeId.tag to OfferNodeId as TlvValueReader<InvoiceRequestTlv>,
+                    // Invoice request part
+                    InvoiceRequestChain.tag to InvoiceRequestChain as TlvValueReader<InvoiceRequestTlv>,
+                    InvoiceRequestAmount.tag to InvoiceRequestAmount as TlvValueReader<InvoiceRequestTlv>,
+                    InvoiceRequestFeatures.tag to InvoiceRequestFeatures as TlvValueReader<InvoiceRequestTlv>,
+                    InvoiceRequestQuantity.tag to InvoiceRequestQuantity as TlvValueReader<InvoiceRequestTlv>,
+                    InvoiceRequestPayerId.tag to InvoiceRequestPayerId as TlvValueReader<InvoiceRequestTlv>,
+                    InvoiceRequestPayerNote.tag to InvoiceRequestPayerNote as TlvValueReader<InvoiceRequestTlv>,
+                    Signature.tag to Signature as TlvValueReader<InvoiceRequestTlv>,
+                )
+            )
+
+            fun decode(s: String): InvoiceRequest {
+                val (prefix, encoded, encoding) = Bech32.decodeBytes(s.lowercase(), true)
+                require(prefix == hrp)
+                require(encoding == Bech32.Encoding.Beck32WithoutChecksum)
+                val tlvs = tlvSerializer.read(encoded)
+                when (val invoiceRequest = validate(tlvs)) {
+                    is Left -> throw IllegalArgumentException(invoiceRequest.value.toString())
+                    is Right -> return invoiceRequest.value
+                }
+            }
+        }
+    }
+
+    data class InvoiceError(val records: TlvStream<InvoiceErrorTlv>) {
+        val error = records.get<Error>()!!.message
+
+        companion object {
+            fun validate(records: TlvStream<InvoiceErrorTlv>): Either<InvalidTlvPayload, InvoiceError> {
+                if (records.get<Error>() == null) return Left(MissingRequiredTlv(5))
+                return Right(InvoiceError(records))
+            }
+        }
+    }
+
+    fun <T : Tlv> rootHash(tlvStream: TlvStream<T>, serializer: TlvStreamSerializer<T>): ByteVector32 {
+        // This encoding/decoding step ensures that the resulting tlvs are ordered.
+        val encoded = serializer.write(tlvStream)
```

Comment:
I think that this can be simplified. In eclair we needed to do this because the `Tlv` trait didn't expose the `tag` as it is entirely handled by `scodec`. But here in `lightning-kmp`, the `Tlv` sealed class exposes the `tag` field, so we can rely on this instead of unnecessarily encoding/decoding.

Reply:
Done